src/main/java/org/broadinstitute/hellbender/engine/FeatureWalker.java

```diff
@@ -20,7 +21,7 @@
  */
 public abstract class FeatureWalker<F extends Feature> extends WalkerBase {
 
-    private FeatureDataSource<F> drivingFeatures;
+    private FeatureInput<F> drivingFeaturesInput;
```

Comment:
@tedsharpe @lbergelson The "historical reason" why we create a special separate additional `FeatureDataSource` for the driving input is twofold:

* There are some tools that need to view the records from the primary input in two ways simultaneously: record-by-record as part of the main traversal, and via windowed queries around the current locus against the `FeatureContext`. This is why we create both a standalone `drivingFeatures` data source, an also add an additional `FeatureInput` for the driving features to `features` (so it will also show up in the `FeatureContext`)

* To get reasonable performance in the above use case, we need different caching settings for the `drivingFeatures` for the main traversal vs. the `FeatureInput` added to the feature sources.

Now, that's the historical reason. The truth is that this only affects certain `VariantWalkers` (and if you read the source for `VariantWalker`, you'll find some comments there on this issue). But `FeatureWalker` was intended to be a generalization of `VariantWalker`, so it would be good if we could keep them consistent if possible.

@tedsharpe Could you briefly explain why you needed to make these changes to `FeatureWalker`? Once I understand your goals here better, I can suggest a way forward. I'm sure we can easily work something out that would require minimal changes on your end!

Reply:
@droazen I've been able to back out almost all my changes to FeatureWalker.  Once I saw that the Tabix reader implemented the famous fake-iterable pattern, I understood what you documented above.
I'm going to rip through the engine classes that I've changed and see what else I can revert.